dask/array/routines.py

```diff
@@ -1531,47 +1541,90 @@ def cov(m, y=None, rowvar=1, bias=0, ddof=None):
 
     # Handles complex arrays too
     m = asarray(m)
+
     if y is None:
-        dtype = np.result_type(m, np.float64)
+        dtype = result_type(m, np.float64)
     else:
         y = asarray(y)
-        dtype = np.result_type(m, y, np.float64)
-    X = array(m, ndmin=2, dtype=dtype)
+        dtype = result_type(m, y, np.float64)
 
-    if X.shape[0] == 1:
-        rowvar = 1
-    if rowvar:
-        N = X.shape[1]
-        axis = 0
-    else:
-        N = X.shape[0]
-        axis = 1
+    if m.ndim > 2:
+        raise ValueError("m has more than 2 dimensions")
+    if y is not None and y.ndim > 2:
+        raise ValueError("y has more than 2 dimensions")
+
+    X = array(m, ndmin=2, dtype=dtype)
 
-    # check ddof
     if ddof is None:
-        if bias == 0:
-            ddof = 1
+        ddof = 1 if bias == 0 else 0
+
+    if not rowvar and m.ndim != 1:
+        X = X.T
+    if X.shape[0] == 0:
+        return array([]).reshape(0, 0)
+    if y is not None:
+        y = array(y, ndmin=2, dtype=dtype)
+        if not rowvar and y.shape[0] != 1:
+            y = y.T
+        X = concatenate((X, y), axis=0)
+
+    # Get the product of frequencies and weights
+    w = None
+    if fweights is not None:
+        fweights = asarray(fweights, dtype=float)
+        if not (fweights == around(fweights)).all():
```

Comment:
Can `fweights` be a large (potentially distribute) array? If so, this round and `.all` might be expensive.

I'd recommend moving this check if it might cause problems.

Reply:
Same for the check on `aweights` below.

Reply:
Both weights are 1-D arrays with the same number of observations as the input 1-D or 2-D array, so yes it could be distributed.